runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -0,0 +1,121 @@
+mod bpf_upgradeable;
+mod builtin;
+pub(crate) mod error;
+
+use {
+    crate::bank::Bank,
+    bpf_upgradeable::BpfUpgradeableConfig,
+    builtin::BuiltinConfig,
+    error::CoreBpfMigrationError,
+    solana_sdk::{
+        account::{Account, AccountSharedData},
+        bpf_loader_upgradeable::{UpgradeableLoaderState, ID as BPF_LOADER_UPGRADEABLE_ID},
+        pubkey::Pubkey,
+    },
+    std::sync::atomic::Ordering::Relaxed,
+};
+
+/// Sets up a Core BPF migration for a built-in program.
+pub enum CoreBpfMigration {
+    Builtin,
+    Ephemeral,
+}
+
+/// Configurations for migrating a built-in program to Core BPF.
+pub struct CoreBpfMigrationConfig {
+    /// The source program ID to replace the builtin with.
+    pub source_program_id: Pubkey,
+    /// The feature gate to trigger the migration to Core BPF.
+    /// Note: This feature gate should never be the same as any builtin's
+    /// `enable_feature_id`. It should always be a feature gate that will be
+    /// activated after the builtin is already enabled.
+    pub feature_id: Pubkey,
+    pub datapoint_name: &'static str,
+}
+
+impl std::fmt::Debug for CoreBpfMigrationConfig {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let mut builder = f.debug_struct("CoreBpfMigrationConfig");
+        builder.field("source_program_id", &self.source_program_id);
+        builder.field("feature_id", &self.feature_id);
+        builder.finish()
+    }
+}
+
+/// Create a new `Account` with a pointer to the target's new data account.
+///
+/// Note the pointer is created manually, as well as the owner and
+/// executable values. The rest is inherited from the source program
+/// account, including the lamports.
+fn create_new_target_program_account(
+    target: &BuiltinConfig,
+    source: &BpfUpgradeableConfig,
+) -> Result<AccountSharedData, CoreBpfMigrationError> {
+    let state = UpgradeableLoaderState::Program {
+        programdata_address: target.program_data_address,
+    };
+    let data = bincode::serialize(&state).map_err(|_| CoreBpfMigrationError::FailedToSerialize)?;
+    let account = Account {
+        data,
+        owner: BPF_LOADER_UPGRADEABLE_ID,
+        executable: true,
```

Comment:
Are these lines necessary? We should only need to change the data, right?

Reply:
If we're totally past `executable` being required, sure that can come out.
Ideally the loader ID _should_ be the correct one, since V2 is deprecated, and you wouldn't make it past the checks if it was a single account.
This was just an extra precaution that I felt had really no drawback, to ensure the program doesn't get bricked.

Reply:
It's definitely still required, I meant more that it can just use whatever's set on the source account, which is what you've done in the recent commits, so this is all good!